pkg/scheduler/flavorassigner/flavorassigner.go

```diff
@@ -218,33 +221,60 @@ func (m FlavorAssignmentMode) String() string {
-	Mode   FlavorAssignmentMode
-	borrow int64
+	Name                  kueue.ResourceFlavorReference
+	Mode                  FlavorAssignmentMode
+	LastAssignedFlavorIdx int
+	borrow                int64
 }
 
 // AssignFlavors assigns flavors for each of the resources requested in each pod set.
 // The result for each pod set is accompanied with reasons why the flavor can't
 // be assigned immediately. Each assigned flavor is accompanied with a
 // FlavorAssignmentMode.
 func AssignFlavors(log logr.Logger, wl *workload.Info, resourceFlavors map[kueue.ResourceFlavorReference]*kueue.ResourceFlavor, cq *cache.ClusterQueue, counts []int32) Assignment {
+	if wl.LastAssignment != nil {
+		if cq.Generation > wl.LastAssignment.ClusterQueueGeneration &&
+			(cq.Cohort == nil || cq.Cohort.Generation > wl.LastAssignment.CohortGeneration) {
+			wl.LastAssignment = nil
+		}
+	}
+
 	if len(counts) == 0 {
-		return assignFlavors(log, wl.TotalRequests, wl.Obj.Spec.PodSets, resourceFlavors, cq)
+		return assignFlavors(log, wl.TotalRequests, wl.Obj.Spec.PodSets, resourceFlavors, cq, wl.LastAssignment)
 	}
 
 	currentResources := make([]workload.PodSetResources, len(wl.TotalRequests))
 	for i := range wl.TotalRequests {
 		currentResources[i] = *wl.TotalRequests[i].ScaledTo(counts[i])
 	}
-	return assignFlavors(log, currentResources, wl.Obj.Spec.PodSets, resourceFlavors, cq)
+	return assignFlavors(log, currentResources, wl.Obj.Spec.PodSets, resourceFlavors, cq, wl.LastAssignment)
 }
 
-func assignFlavors(log logr.Logger, requests []workload.PodSetResources, podSets []kueue.PodSet, resourceFlavors map[kueue.ResourceFlavorReference]*kueue.ResourceFlavor, cq *cache.ClusterQueue) Assignment {
-	assignment := Assignment{
-		TotalBorrow: make(cache.FlavorResourceQuantities),
-		PodSets:     make([]PodSetAssignment, 0, len(requests)),
-		Usage:       make(cache.FlavorResourceQuantities),
+func assignFlavors(log logr.Logger, requests []workload.PodSetResources, podSets []kueue.PodSet, resourceFlavors map[kueue.ResourceFlavorReference]*kueue.ResourceFlavor, cq *cache.ClusterQueue, lastAssignment *workload.AssigmentClusterQueueState) Assignment {
+	var assignment Assignment
+	if lastAssignment != nil {
+		assignment = Assignment{
+			TotalBorrow: make(workload.FlavorResourceQuantities),
+			PodSets:     make([]PodSetAssignment, 0, len(requests)),
+			LastState:   *lastAssignment,
+			Usage:       make(workload.FlavorResourceQuantities),
+		}
+	} else {
+		assignment = Assignment{
+			TotalBorrow: make(workload.FlavorResourceQuantities),
+			PodSets:     make([]PodSetAssignment, 0, len(requests)),
+			LastState: workload.AssigmentClusterQueueState{
+				LastAssignedFlavorIdx:  make([]map[corev1.ResourceName]int, 0),
+				CohortGeneration:       0,
+				ClusterQueueGeneration: cq.Generation,
+			},
+			Usage: make(workload.FlavorResourceQuantities),
+		}
+		if cq.Cohort != nil {
+			assignment.LastState.CohortGeneration = cq.Cohort.Generation
+		}
 	}
```

Comment:
why not keep the code as is, and add
```golang
if lastAssignment != nil {
  assigment.LastState = ...
} else {
  assignment.LastState = ...
}
```

it's less repeated code and more clear what is the actual difference.

Reply:
Fixed